src/lib/sql/sql_translator.cpp

```diff
@@ -403,15 +410,103 @@ std::shared_ptr<AbstractLQPNode> SQLTranslator::_translate_join(const hsql::Join
   return join_node;
 }
 
+LQPInputSide SQLTranslator::_get_side(const std::shared_ptr<AbstractLQPNode>& left_node,
+                                      const std::shared_ptr<AbstractLQPNode>& right_node, const hsql::Expr* expr) {
+  DebugAssert(expr && expr->type == hsql::kExprColumnRef, "only applicable to columnar expressions");
+  const auto qualified = HSQLExprTranslator::to_qualified_column_name(*expr);
+  if (left_node->find_column(qualified)) {
+    return LQPInputSide::Left;
+  }
+
+  if (right_node->find_column(qualified)) {
+    return LQPInputSide::Right;
+  }
+
+  Fail("Column not found: " + qualified.as_string());
+}
+
+void SQLTranslator::_get_sides_from_operator_expressions(std::vector<const hsql::Expr*> operators,
+                                                         std::function<LQPInputSide(const hsql::Expr*)> get_side,
+                                                         std::vector<const hsql::Expr*>& both,
+                                                         std::vector<const hsql::Expr*>& left,
+                                                         std::vector<const hsql::Expr*>& right) {
+  for (const auto* op : operators) {
+    const bool left_is_column = op->expr && op->expr->type == hsql::kExprColumnRef;
+    const bool right_is_column = op->expr2 && op->expr2->type == hsql::kExprColumnRef;
+
+    if (left_is_column && right_is_column) {
+      auto left_column_side = get_side(op->expr);
+      auto right_column_side = get_side(op->expr2);
+      if (left_column_side == right_column_side) {
+        if (left_column_side == LQPInputSide::Left) {
+          left.push_back(op);
+        } else {
+          right.push_back(op);
+        }
+      } else {
+        both.push_back(op);
+      }
+    } else {
+      if (left_is_column) {
+        if (get_side(op->expr) == LQPInputSide::Left) {
+          left.push_back(op);
+        } else {
+          right.push_back(op);
+        }
+      } else if (right_is_column) {
+        if (get_side(op->expr2) == LQPInputSide::Left) {
+          left.push_back(op);
+        } else {
+          right.push_back(op);
+        }
+      }
+    }
+  }
+}
+
+void SQLTranslator::_apply_predicate_expressions(const std::shared_ptr<AbstractLQPNode>& node, const LQPInputSide side,
+                                                 const std::vector<const hsql::Expr*>& conditions) {
+  if (conditions.empty()) {
+    return;
+  }
+
+  std::vector<std::shared_ptr<AbstractLQPNode>> predicates;
+  for (const auto* condition : conditions) {
+    predicates.push_back(_translate_predicate(
+        *condition, false, [&node](const auto& expr) { return HSQLExprTranslator::to_column_reference(expr, node); },
+        node));
+  }
+
+  for (size_t index = 0; index < predicates.size() - 1; ++index) {
+    predicates[index]->set_left_input(predicates[index + 1]);
+  }
+
+  predicates.back()->set_left_input(node->input(side));
+  node->set_input(side, predicates.front());
```

Comment:
You should check that `node` had an input on that side before.

Reply:
Why exactly? Of particular importance for this method is the fact that the predicates appear in the subtree of the given side. In presence of other nodes in the subtree, the predicates should appear in between of the root node and its previous input. Currently I do not see good reason to restrict the routine to nodes with existing inputs.

Reply:
Right now, I could call your method to add some predicates on the right side of a PredicateNode or the left side of a GetTable. Neither make sense.